mdata/cache/chunk_cache.go

```diff
@@ -0,0 +1,96 @@
+package cache
+
+import (
+	"flag"
+	"sync"
+
+	"github.com/raintank/metrictank/iter"
+	"github.com/rakyll/globalconf"
+)
+
+var (
+	maxSize uint64
+)
+
+func ConfigSetup() {
+	flags := flag.NewFlagSet("chunk-cache", flag.ExitOnError)
+	// (1024 ^ 3) * 4 = 4294967296
+	flags.Uint64Var(&maxSize, "max-size", 4294967296, "Maximum size of chunk cache in bytes")
+	globalconf.Register("chunk-cache", flags)
+}
+
+type CacheChunk struct {
+	Ts      uint32
+	Next    uint32
+	Prev    uint32
+	itergen iter.IterGen
+}
+
+type CCacheMetric struct {
+	sync.RWMutex
+	chunks map[uint32]CacheChunk
+}
+
+type CCache struct {
+	sync.RWMutex
+	metricCache map[string]*CCacheMetric
+}
+
+func NewChunkCache() *CCache {
+	return &CCache{
+		metricCache: make(map[string]*CCacheMetric),
+	}
+}
+
+func (c *CCache) Add(metric string, prev uint32, itergen iter.IterGen) error {
+	c.RLock()
+	if _, ok := c.metricCache[metric]; !ok {
+		c.RUnlock()
+
+		c.Lock()
```

Comment:
isn't this a race condition? in between the RUnlock and the Lock another thread could have added the CCacheMetric and perhaps even added chunks to it.
Probably better to just do the ok check and the adding to the map in 1 critical section